### PR TITLE
fix point key next's semantic meaning

### PR DESCRIPTION
**Pull request by zhexuany**

### CODE REVIEW

File: core/src/main/scala/com/pingcap/tispark/examples/TiExtensionsExample.scala [hunk 1]
```diff
@@ -28,16 +28,36 @@ object TiExtensionsExample {
...
+      .sql(
+        s"""
+           |select stat_date,device_id,city_id as ucity_id,cid_id,diary_service_id
+           |from jerry_prod.data_feed_exposure
+           |where cid_type = "diary" limit 6
+         """.stripMargin
+      )
+      .show(6)
+   */
+
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/examples/TiExtensionsExample.scala]: do not need to change this file

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala [hunk 2]
```diff
@@ -254,6 +254,20 @@ class IssueTestSuite extends BaseTiSparkSuite {
...
+  test("index double scan with predicate") {
+    tidbStmt.execute("drop table if exists test_index")
+    tidbStmt.execute(
+      "create table test_index(id bigint(20), c1 text default null, c2 int, c3 int, c4 int, KEY idx_c1(c1(10)))"
+    )
+    tidbStmt.execute("insert into test_index values(1, 'aairy', 10, 20, 30)")
+    tidbStmt.execute("insert into test_index values(1, 'dairy', 10, 20, 30)")
+    tidbStmt.execute("insert into test_index values(1, 'zairy', 10, 20, 30)")
+    refreshConnections() // refresh since we need to load data again
+    explainAndRunTest("select c2 from test_index where c1 > 'dairy'")
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: maybe use `judge` ?

File: tikv-client/src/main/java/com/pingcap/tikv/key/Key.java [hunk 3]
```diff
@@ -116,6 +116,15 @@ public Key next() {
     return toRawKey(Arrays.copyOf(value, value.length + 1));
   }
 
+  /**
+   * nextPrefix key will be key with next available rid.
```
Comment (marsishandsome) [inline tikv-client/src/main/java/com/pingcap/tikv/key/Key.java]: it will be better if you can provide some examples

File: tikv-client/src/main/java/com/pingcap/tikv/key/IndexScanKeyRangeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import static com.pingcap.tikv.util.KeyRangeUtils.makeCoprocRange;
+
+import com.google.common.collect.BoundType;
+import com.google.common.collect.Range;
+import com.pingcap.tikv.meta.TiIndexInfo;
+import com.pingcap.tikv.predicates.IndexRange;
+import org.tikv.kvproto.Coprocessor.KeyRange;
+
+public class IndexScanKeyRangeBuilder {
```
Comment (marsishandsome) [inline tikv-client/src/main/java/com/pingcap/tikv/key/IndexScanKeyRangeBuilder.java]: pls add some comments

File: core/src/main/scala/com/pingcap/tispark/examples/TiExtensionsExample.scala (hunk 1 above)
Comment (zhexuany) [reply core/src/main/scala/com/pingcap/tispark/examples/TiExtensionsExample.scala]: done

File: tikv-client/src/main/java/com/pingcap/tikv/key/Key.java (hunk 3 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/key/Key.java]: done

**Comment by marsishandsome** [comment]
@birdstorm PTAL

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java [hunk 5]
```diff
@@ -369,61 +369,62 @@ else if (col.getColumnInfo().isPrimaryKey() && tableInfo.isPkHandle()) {
       }
 
       dagRequestBuilder.addExecutors(executorBuilder.setTblScan(tblScanBuilder));
-    }
 
-    // clear executorBuilder
```
Comment (zhexuany) [inline tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java]: changes in this file, just move selection, limit executor building process into table scan.

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala (hunk 2 above)
Comment (zhexuany) [reply core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: done

File: tikv-client/src/main/java/com/pingcap/tikv/key/Key.java [hunk 6]
```diff
@@ -116,6 +116,16 @@ public Key next() {
     return toRawKey(Arrays.copyOf(value, value.length + 1));
   }
 
+  /**
+   * nextPrefix key will be key with next available rid. For example, if the current key is
+   * prefix_rid, after calling this method, the return value should be prefix_rid+1
+   *
+   * @return a new key current rid+1.
+   */
+  public Key nextPrefix() {
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/key/Key.java]: any tests for nextPrefix() ?

File: tikv-client/src/main/java/com/pingcap/tikv/key/IndexScanKeyRangeBuilder.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
+package com.pingcap.tikv.key;
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/key/IndexScanKeyRangeBuilder.java]: copyright

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala [hunk 8]
```diff
@@ -254,6 +254,20 @@ class IssueTestSuite extends BaseTiSparkSuite {
     judge("select cast(count(1) as char(20)) from `tmp_empty_tbl`")
   }
 
+  test("index double scan with predicate") {
```
Comment (birdstorm) [inline core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: i suggest we add these tests to expression/index

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: we also need to test aggregations and duplicate index columns.

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala [hunk 9]
```diff
@@ -254,6 +254,20 @@ class IssueTestSuite extends BaseTiSparkSuite {
...
+    tidbStmt.execute("drop table if exists test_index")
+    tidbStmt.execute(
+      "create table test_index(id bigint(20), c1 text default null, c2 int, c3 int, c4 int, KEY idx_c1(c1(10)))"
+    )
+    tidbStmt.execute("insert into test_index values(1, 'aairy', 10, 20, 30)")
```
Comment (birdstorm) [inline core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: can you make sure these tests go into double read logic?

File: tikv-client/src/main/java/com/pingcap/tikv/key/Key.java (hunk 6 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/key/Key.java]: DONE

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala (hunk 8 above)
Comment (zhexuany) [reply core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: DONE
